### DISCUSSION TITLE
Cannot connect over USB to virtual serial port of Crazyflie 2.1

### DISCUSSION THREAD

**Discussion by wiedenmayerjulian** [bitcraze/discussions#1636] [General]
I tried multiple times to connect to the /dev/ttyACM0 USB device to get the console logs (without cfclient).

It seems not to work. I tried different baudrates and so on. 

Is it possible to read the console output from USB without Crazyradio?

**Comment by Bananazooka** [comment]
Hi, did you set up the USB permissions? You can find instructions [here](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/installation/usb_permissions/).

**Reply by wiedenmayerjulian** [reply]
Hi, thank you very much for your reply! I tried everything as root user. Should i set up USB permissions and retry as normal user?

**Reply by knmcguire** [reply]
yes that is what we advise our users to do, also if you use WSL2

**Comment by NageshEranki** [comment]
Hi 

1. Could you try connecting via cfclient? If you are unable to do so then you still have USB issues. If you are able to connect to the drone from cfclient, then the issue may be with your script.
2. Could you share the python script you are using to log console output?
3. Could you describe your platform details. For e.g., operating system, whether you are using a virtual machine, what python environment you are using, etc?

Hope this helps!

**Reply by wiedenmayerjulian** [reply]
Hi, thank you for your reply.

1. I could connect with cfclient 2024.10, Crazyradio 2.0 dongle and read the logs in the console tab. Which script do you mean is affected then, when it connects via cfclient?
2. I installed cflib 0.1.27 via pip. Therefore it must be this: https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/console.py 
3. Windows 10 64 Bit with Python 3.12.6. 
To access USB via /dev/ttyACM0 I installed Ubuntu 22.04.5 LTS in WSL Version 2. Python 3 is also installed in Ubuntu.

On both systems is pyusb installed. In Windows the windows usb driver is installed:

<img width="266" alt="image" src="https://github.com/user-attachments/assets/f5fc16d7-b790-4f9f-909d-c5b6d7fcd5fe">


For WSL USB Support I installed the USBIPD-WIN project (https://learn.microsoft.com/en-us/windows/wsl/connect-usb) and attached the "Crazyflie USB device" to WSL:
```PS C:\Windows\system32> usbipd list
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
1-2    0483:5740  Crazyflie 2.x                                                 Not shared
1-3    1915:7777  Crazyradio 2.0                                                Not shared
1-8    04f3:0c11  ELAN WBF Fingerprint Sensor                                   Not shared

...
PS C:\Windows\system32> usbipd bind -b 1-2
...
PS C:\Windows\system32> usbipd attach --wsl --busid 1-2
usbipd: info: Using WSL distribution 'Ubuntu' to attach; the device will be available in all WSL 2 distributions.
usbipd: info: Using IP address 172.28.64.1 to reach the host.
```

When the USB device is attached, /dev/ttyACM0 is shown in Ubuntu. 

```
root@AORUS-17GXC-1:~# ls /dev/tty
tty      tty13    tty19    tty24    tty3     tty35    tty40    tty46    tty51    tty57    tty62    ttyS0
tty0     tty14    tty2     tty25    tty30    tty36    tty41    tty47    tty52    tty58    tty63    ttyS1
tty1     tty15    tty20    tty26    tty31    tty37    tty42    tty48    tty53    tty59    tty7     ttyS2
tty10    tty16    tty21    tty27    tty32    tty38    tty43    tty49    tty54    tty6     tty8     ttyS3
tty11    tty17    tty22    tty28    tty33    tty39    tty44    tty5     tty55    tty60    tty9
tty12    tty18    tty23    tty29    tty34    tty4     tty45    tty50    tty56    tty61    ttyACM0
```

There i tried to open /dev/ttyACM0:
```root@AORUS-17GXC-1:/# picocom --b 115200 /dev/ttyACM0
picocom v3.1

port is        : /dev/ttyACM0
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        :
omap is        :
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready
```



I also tried minicom:

```
root@AORUS-17GXC-1:~# minicom -D /dev/ttyACM0 -b 115200
```

But i cannot get any output in terminal.

```
Welcome to minicom 2.8

OPTIONS: I18n
Port /dev/ttyACM0, 21:45:05

Press CTRL-A Z for help on special keys



CTRL-A Z for help | 115200 8N1 | NOR | Minicom 2.8 | VT102 | Offline | ttyACM0
```

What can I do to fix that?

**Comment by knmcguire** [comment]
Mind that the Crazyflie python library has a script example to get console output on the terminal. You'll need to change the URI of the crazyflie to usb://0

https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/console/console.py

**Reply by wiedenmayerjulian** [reply]
Thank you very much! That's also a good opportunity to access console output. But is there a "bare metal"-way over minicom /dev/ttyACM0 or similar?

**Comment by wiedenmayerjulian** [comment]
In addition to this: Is it possible to access console tty like UART does but instead of UART using USB?

**Reply by knmcguire** [reply]
Yes there is actually. The crazyflie library does allow uart connection, but you'll probably need to enable something in the firmware build.

Here is documentation that can point you in the right direction:
- https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/
- https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/

Let us know if you have any more questions
